### ISSUE TITLE
Add option to set encryption key for storage

### ISSUE THREAD

**Issue by wburningham** [gomods/athens#1539]
**Is your feature request related to a problem? Please describe.**
I have two use cases where data needs to be:
1. "encrypted at rest" (for file storage backend)
2. "encrypted in the runtime/process" even though TLS encrypts HTTP traffic and even though some storage providers may provide encryption at rest (ex Amazon's S3), storage providers still have my plaintext data in their servers and I just trusting providers is not a sufficient level of security

**Describe the solution you'd like**
Provide an `ATHENS_STORAGE_ENCRYPTION_KEY` option that if set will wrap a `storage.Backend` and handle encrypting and decrypting of data.

**Describe alternatives you've considered**
Write an encryption backend pkg that wraps `storage.Backend` and created my own proxy app that uses most of the `gomods/athens/pkg/*` internals. I've done this, but I think writing a PR back to `gomods/athens` with the encryption would be a better solution.

**Comment by arschles** [comment]
@wburningham for the 2nd requirement, could Athens encrypt before sending to S3 and decrypt in-memory after receiving data?

Also, do you have code that you can share from your own proxy app? We can use that and work together to submit a PR back to Athens if you'd like

**Comment by wburningham** [comment]
The code that I had from my own app was already in it's own package so I just submitted a PR with what it could look like.

**Comment by marwan-at-work** [comment]
I think this is a perfect use case for https://github.com/gomods/athens/pull/1353

Maybe it's worth revisiting so that we don't inflate our backend implementations. 

Admittedly, encrypting/decrypting modules within the Athens codebase sounds pretty damn cool ;) 

But I worry about the following:

If we were to claim Athens has encryption within its codebase, then people will trust it and cryptographers will put it under scrutiny.  

None of the maintainers (AFAIK), or cryptography experts to maintain and develop proper encryption. Therefore, we would need a cryptography champion who is actively developing and maintaining this backend implementation if we are to have Athens claim e2e support :)

**Comment by arschles** [comment]
>I think this is a perfect use case for #1353

As it happens, I was just commenting on an unrelated issue where the author could use something like #1353. Maybe it's time 😄 

If we do any encryption work, I'm all for putting a big disclaimer that we're not experts, until we get someone to audit and/or champion the implementation

**Comment by marwan-at-work** [comment]
@arschles ideally I'd go for 1353 first, it would also help keep the Athens code base smaller and easier to maintainer since we already have a big number of implementations. I definitely don't wanna block anyone tho in case this is absolutely needed 👍

**Comment by arschles** [comment]
Agreed on #1353. Looks like that PR hasn't had any activity for a while, so I'm gonna open a new draft with a simple implementation and we can start there.
